### ISSUE TITLE
[BUG]: Fallback handler could not load library issue on Mac(2018.4.36)

### ISSUE THREAD

**Issue by Yuuon** [Unity-Technologies/com.unity.webrtc#636]
### Package version

2.4.0-exp.4

### Environment

```markdown
* OS:macOS BigSur 11.5.1(x64) / macOS monterey 12.2.1(Silicon)
* Unity version: 2018.4.36
```

### Steps To Reproduce

1. In Unity Editor, change the built plugin file extend name from .dylib to .bundle file so Unity Editor could regconize it and modify settings.
2. Set the plugin target platform to MacOS and x64
3. Build our project, do codesign and notarize, and get the application, run it.
4. Run a HIFI(voice chat plugin which using webrtc functions) test
5. In Log, can find a failure on read plugin file and finally with a notification of DLLNotFoundException, and all later function broken with null reference error. Related logs see below:

> Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.dylib
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.so
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.bundle
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/webrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.dylib
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.so
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.bundle
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.dylib
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.so
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.bundle
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/webrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.dylib
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.so
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.bundle
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.dylib
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.so
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.bundle
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/webrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.dylib
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.so
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc.bundle
Fallback handler could not load library /private/var/folders/dj/z3cz416n7ml7lg4t9m7946tc0000gn/T/AppTranslocation/2F4F279E-1A16-414A-AF75-FC17BF175DB4/d/Sinespace.app/Contents/Frameworks/MonoEmbedRuntime/osx/libwebrtc
DllNotFoundException: webrtc
  at (wrapper managed-to-native) Unity.WebRTC.NativeMethods.RegisterDebugLog(Unity.WebRTC.DelegateDebugLog,bool,Unity.WebRTC.NativeLoggingSeverity)
  at Unity.WebRTC.WebRTC.Initialize (Unity.WebRTC.EncoderType type, System.Boolean limitTextureSize, System.Boolean forTest, System.Boolean enableNativeLog, Unity.WebRTC.NativeLoggingSeverity nativeLoggingSeverity) [0x0005d] in <0750a01427d34a8dab297b75c23a2982>:0 
  at Unity.WebRTC.WebRTC.Initialize (Unity.WebRTC.EncoderType type, System.Boolean limitTextureSize, System.Boolean enableNativeLog, Unity.WebRTC.NativeLoggingSeverity nativeLoggingSeverity) [0x00012] in <0750a01427d34a8dab297b75c23a2982>:0 
  at Ravi.RaviUtil.InitializeWebRTC () [0x0000d] in <34d28c1da819455bb43fd68c3ffbd88b>:0 
  at HiFi.HiFiCommunicator.Awake () [0x00016] in <34d28c1da819455bb43fd68c3ffbd88b>:0 
UnityEngine.GameObject:Internal_AddComponentWithType(Type)
UnityEngine.GameObject:AddComponent(Type)
UnityEngine.GameObject:AddComponent()
HiFiClient:Connect(String)
HiFiClient:PerformJoinRegion(String)
VoiceToggleFuncControlBaseInternal:JoinVoice(String)
VoiceToggleFuncControlBaseInternal:ProcessZoneChange()
MediaManagerBaseInternal:Update()

### Current Behavior

In Mac built, because of the failure of reading DLL, all functions related is broken.
Both Intel and Silicon platform is tested and result is same.
The build is built by mono. We didn't test IL2CPP yet on desktop platform.
However, when test in Unity Editor, it seems fine.
Also there is no issue on Windows Build.

### Expected Behavior

As Windows version works fine, we hope Mac build could get the same result.
Is it really limited that the webrtc could only used on Unity 2019.4 or above?

### Anything else?

Or maybe we've done something wrong on generated the dylib file?
I upload the libwebrtc file we made to Dropbox: [https://www.dropbox.com/s/zaxmmpskzut316n/libwebrtc.bundle?dl=0](url)

**Comment by Yuuon** [comment]
It seems that Unity 2018 doesn't have support for dylib files, so are there any instructions to make a bundle build from the webrtc source code? I'm not familar with Apple things at all...

**Comment by karasusan** [comment]
Could you check it?
#568

**Comment by Yuuon** [comment]
> Could you check it? #568

It seems like an inverse situation, Mac Silicon will cause crash issue so you've changed bundle to dylib, however, this drops support for Unity 2018. Currently what I've made clear is Unity 2018 doesn't recognize dylib file so when build the application, the file won't get include automatically, also simply change the filename from .dylib to .bundle(or anything else Unity 2018 could recognize) can let the Unity Editor show options in inspector, but the built application still have dllNotFoundException.
From Unity 2019 there will be native support for dylib files but upgrade the whole project is not an option for now.

What I've seen that the only thing I can do is build the plugin into the bundles by myself, but actually I have no idea where to start, probably makefile?

**Comment by Yuuon** [comment]
So bascically I'm following this part: [https://github.com/Unity-Technologies/com.unity.webrtc/tree/release/2.3.3/Plugin~](url)
But met question that dependancy is missing:

> % cmake . -G Xcode -B build
CMake Error at /usr/local/Cellar/cmake/3.22.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find WebRTC (missing: WEBRTC_LIBRARY)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.22.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindWebRTC.cmake:51 (find_package_handle_standard_args)
  CMakeLists.txt:65 (find_package)

I'm following what the instruction saying, but using the M96 webrtc-mac release, extract the files of libwebrtc(includes and .a lib file) and put them under the webrtc folder in Plugin~.
Anything I did wrong or I should manually built the libwebrtc too?

**Comment by karasusan** [comment]
Looks like downloading or extracting libwebrtc fails.
And we haven't finished to integrate M96. Please try with M92.

**Comment by karasusan** [comment]
@Yuuon 
Basically, we supports Unity LTS versions. Unfortunately, Unity 2018 is already removed from the support list.
https://blog.unity.com/en/technology/the-2-year-support-for-the-first-lts-release-2017-4-ends-in-april-2020

**Comment by Yuuon** [comment]
Eh... I've tried the old way but I still get the dylib file, so it's still the makefile which related to the output? Or it is already decided in that .a file?
I know 2018 is not LTS anymore but the cost of upgrade project is not an easy thing. We do have plan of upgrade to 2020 and we are preparing for it, however it's for next release which is still months ahead, so currently still need make some effort on 2018.
Personally I've seen is make the plugin into bundle structure as it used to be. I'm not sure how hard it is but it should be possible as you've done this before, also if there are some instructions I can handle it myself, i.e. I'm looking back the version 2.3.3 which has a bundle output. If you can tell me which files I should look into or build natively, that would be appreciated.

**Comment by karasusan** [comment]
@Yuuon
Maybe you can build the bundle file if you revert the changeset #568.

**Comment by Yuuon** [comment]
Yes I've found where the difference is, it's in a CMakeFileList, it removed the bundle settings for mac os. Successfully get the bundle file now.

**Comment by Yuuon** [comment]
Sorry need to reopen the issue, met a new issue related to the libwebrtc version.
So yesterday with the M92 pack I successfully get the bundle but when put it in the application, it report an InvalidCastException so obviously there are something changed between versions. Using M96 with 2.4.0-exp5 is still getting error when call cmake: cmake . -G Xcode -B build

> CMake Error at /usr/local/Cellar/cmake/3.22.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find WebRTC (missing: WEBRTC_LIBRARY WEBRTC_LIBRARY_DEBUG
  WEBRTC_LIBRARY_RELEASE)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.22.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindWebRTC.cmake:51 (find_package_handle_standard_args)
  CMakeLists.txt:65 (find_package)

As you've mentioned the M96 pack may have some thing lost?
So where can I get the latest and complete libwebrtc?

**Comment by karasusan** [comment]
https://github.com/Unity-Technologies/com.unity.webrtc/issues/636#issuecomment-1054013031

**Comment by Yuuon** [comment]
M92 leads this error:

> EntryPointNotFoundException: ContextRegisterAudioReceiveCallback
  at (wrapper managed-to-native) Unity.WebRTC.NativeMethods.ContextRegisterAudioReceiveCallback(intptr,intptr,Unity.WebRTC.DelegateAudioReceive)
  at Unity.WebRTC.Context.AudioTrackRegisterAudioReceiveCallback (System.IntPtr track, Unity.WebRTC.DelegateAudioReceive callback) [0x00000] in <3682c3c66c92460397533fefe06ca984>:0 
  at Unity.WebRTC.AudioStreamTrack..ctor (System.IntPtr ptr) [0x0000c] in <3682c3c66c92460397533fefe06ca984>:0 
  at Unity.WebRTC.AudioStreamTrack..ctor (System.String label, Unity.WebRTC.AudioTrackSource source) [0x00012] in <3682c3c66c92460397533fefe06ca984>:0 
  at Unity.WebRTC.AudioStreamTrack..ctor (UnityEngine.AudioSource source) [0x00019] in <3682c3c66c92460397533fefe06ca984>:0 
  at Ravi.RaviSession.CaptureInputAudio () [0x000bb] in <2f528e65de9e4fa99bdf734d0165a84c>:0 
  at Ravi.RaviSession.Start () [0x0000b] in <2f528e65de9e4fa99bdf734d0165a84c>:0 

which related with file com.unity.webrtc/Runtime/Scripts/WebRTC.cs:
[DllImport(WebRTC.Lib)]
public static extern void ContextRegisterAudioReceiveCallback(IntPtr context, IntPtr track, DelegateAudioReceive callback);

which is added in #453 , on 5 July 2021, and M92 is published on 10 July 2021 so from the timeline it seems should contain this function but I had no way to verify.
Can you help me to check what I'm missing here?

**Comment by karasusan** [comment]
The exception means that the function is not existed in the native plugin.
`ContextRegisterAudioReceiveCallback` is not existed in this latest code. Maybe you are building with old one.
https://github.com/Unity-Technologies/com.unity.webrtc/blob/develop/Plugin~/WebRTCPlugin/WebRTCPlugin.cpp

**Comment by Yuuon** [comment]
I've tried again and following the steps below:
1. Download the source code of the latest release 2.4.0-exp.5
2. Download the libwebrtc file of version M92 for mac: webrtc-mac.zip
3. Copy them to a Mac machine and extract them, put libwebrtc contents into Plugin~/webrtc/ folder
4. Modify Plugin~/WebRTCPlugin/CMakeFileLists.txt, adding bundle option as the same file from 2.3.3 package.
5. Back to Plugin~/ and run "cmake . -G Xcode -B build" then "cmake --build build --config Release --target WebRTCPlugin"
6. In Runtime/ folder, find the new built webrtc.bundle plugin
7. In the appliction I built with the previous webrtc.bundle I made yesterday, in Contents/Plugins/, replace the old one with the new one.
8. Run the application and go for audio test, then check the unity log of the application.

Actual result: reporting EntryPointNotFoundException: ContextRegisterAudioReceiveCallback
I'm a bit confusing now cuz in the latest code the building script is also using M92 package. I've just noticed in BuildScripts~/ folder that there is a script for build libwebrtc -- should I manually make this too if I want to use 2.4.0-exp.5 to make the bundle?
